[Mobile] Update the bottom sheet header

enejb opened the pull request:
## Description
This PR Refactors the Navigation Header Component so that it is more compostable, easier to understand and updates the usage of the component. 

## How has this been tested?
1. Load up this PR on a mobile simulator.
2. Navigate to different Bottoms Sheets. Notice that they look the same as expected. 
Bottoms sheets to test out. 

- Button Block. 
  - Button Width.
  - Link Settings.
- Image block
  - Alt text
  - Size Picker
- Cover Block.
  - Edit Focal Point
  - Color Picker
- Columns Block
  - Columns Selector
- Paragraph Block
  - Font Site Settings. 
  - Colour Settings
  - Gradient Settings
- Help Section. 


## Screenshots <!-- if applicable -->
The Bottom Sheet Header should look exactly the same as before. 
Something like this.
<img width="432" alt="Screen Shot 2021-08-25 at 3 02 27 PM" src="https://user-images.githubusercontent.com/115071/130870397-dbeee8ca-f28f-4fb1-ae0d-0cd74d64cec6.png">


## Types of changes
Refactor

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

enejb committed:
Add new Header bottomSheet component

enejb committed:
Include the new Header component in buttomSheet

enejb committed:
Update NavigationHeader to use the new Header component for backwards compatibility

enejb committed:
Update all instances of NavigationHeader with Header

SiobhyB commented on packages/components/src/mobile/bottom-sheet/index.native.js:
```diff
@@ -600,6 +601,7 @@ ThemedBottomSheet.Button = Button;
 ThemedBottomSheet.Cell = Cell;
 ThemedBottomSheet.SubSheet = BottomSheetSubSheet;
 ThemedBottomSheet.NavigationHeader = NavigationHeader;
```
As far as I can see, the `NavigationHeader` component is no longer in use anywhere and removing this reference doesn't have any effect on the BottomSheet's header component. Is there a reason for keeping the `bottom-sheet/navigation-header.native.js` file and references to the component?

enejb replied:
I think we can remove it. Good catch!

SiobhyB commented on packages/components/src/mobile/bottom-sheet/header/README.md:
```diff
@@ -0,0 +1,91 @@
+Header components is meant to be used to compose the header inside the BottomSheet.
+
+### Usage
+See 
```
```suggestion
```

It looks like this can either be removed or perhaps there was a missing referencing?

enejb replied:
Thanks for the suggestion @SiobhyB The documentation section got converted updated in https://github.com/WordPress/gutenberg/pull/34339.

SiobhyB commented on packages/components/src/mobile/bottom-sheet/header/README.md:
```diff
@@ -0,0 +1,91 @@
+# Header
+
+Header components is meant to be used to compose the header inside the BottomSheet.
```
```suggestion
`BottomSheet.Header` should be used to add a header to the top of a `BottomSheet` component. It makes several other components available, which can then be used to compose the header's content.
```
It took me a bit to fully process how this component works, so think adding a bit more detail to the intro might help with clarity. WDYT?

SiobhyB commented on packages/components/src/mobile/bottom-sheet/header/README.md:
```diff
@@ -0,0 +1,91 @@
+The apply button is used to apply settings of the bottom sheet control.
+
+##### onPress 
+use to pass a call back once the ApplyButton is clicked. 
```
```suggestion
Use to pass a call back once the ApplyButton is clicked. 
```

SiobhyB commented on packages/components/src/mobile/bottom-sheet/header/README.md:
```diff
@@ -0,0 +1,91 @@
+The back button is used to send the user back to the previous sheet. 
+
+##### onPress 
+use to pass a call back once the BackButton is clicked. 
```
```suggestion
Use to pass a call back once the BackButton is clicked. 
```

SiobhyB commented on packages/components/src/mobile/bottom-sheet/header/README.md:
```diff
@@ -0,0 +1,91 @@
+The cancel button is used to send the user back to the previous sheet. Closes the sheet. Use this if you are using a full screen bottom sheet. 
+
+##### onPress 
+use to pass a call back once the CancelButton is clicked. 
```
```suggestion
Use to pass a call back once the CancelButton is clicked. 
```

SiobhyB reviewed (approved):
I tested the following in the demo app for both iOS and Android, including both light and dark mode, and confirm that the header looks visually the same as before the changes:

<hr />

- [x] **Button Block**
  - [x] Button Width
  - [x] Link Settings

- [x] **Image block**
  - [x] Alt text

- [x] **Cover Block**
  - [x] Edit Focal Point
  - [x] Color Picker

- [x] **Paragraph Block**
  - [x] Font Site Settings.
  - [x] Colour Settings

<hr />

I made a couple of small suggestions for the README and also asked a question about the `navigation-header-native.js` file (i.e. whether there's a reason it can't be deleted). Other than that, all the code makes sense to me and I agree the updated component is more readable. :) As I couldn't spot any obvious big issues or blockers, I've approved from my side.

dcalhoun commented on packages/components/src/mobile/bottom-sheet/header/action-button.native.js:
```diff
@@ -0,0 +1,29 @@
+	children,
+} ) {
+	return (
+		<TouchableWithoutFeedback
```
I have often wondered why `TouchableWithoutFeedback` without feedback is used for many of the interactive UI elements within this project. It would likely improve the usability of this component if we leverage `TouchableOpacity`, `TouchableHighlight`, or a custom style as a feedback signal when this element is pressed. I.e. currently, when this button is pressed, there is no visual feedback to signal that the touch was registered by the UI. This can lead to a confusing experience. 

WDYT?

enejb replied:
I think adding some touch feedback I think is a great idea. I think that could be tacked in a different PR. Mostly because this one is already pretty big and the change could require some discussion and input from design folks.

mkevins replied:
I agree about `TouchableWithoutFeedback`, and also agree that it may be best to tackle in another PR. I think auditing project-wide usage of Touchable components (and making the appropriate changes) would be a fairly big PR even as a stand-alone PR (and definitely a good idea, too). :smiley:

dcalhoun replied:
Agreed. Modifying the interaction design is better suited for a separate PR. We can resolve this conversation.

dcalhoun commented on packages/components/src/mobile/bottom-sheet/header/back-button.native.js:
```diff
@@ -0,0 +1,99 @@
+				accessibilityLabel={ __( 'Go back' ) }
+				accessibilityHint={ __(
+					'Navigates to the previous content sheet'
+				) }
```
For contexts like "cancel," I'm not sure if this a11y label and hint most accurately describes the intent. E.g. a label of "Cancel changes" may be more accurate than "Go back." We may need to avoid abstracting this or provide a mechanism for altering it. WDYT?

enejb replied:
That makes sense to me. During my refactor I made some changes to implement different hints and labels but then I didn't end up doing it mostly because I didn't want to change what is currently there. (This PR was getting pretty big) 
I can take that change in a different follow-up PR :) to this one.

dcalhoun replied:
Ah, OK. I agree it makes sense to keep this PR focused on refactoring and capture any modifications to accessibility properties in a separate PR. We can resolve this conversation.

dcalhoun commented on packages/components/src/mobile/bottom-sheet/header/back-button.native.js:
```diff
@@ -0,0 +1,99 @@
+
+function CloseButton( { onPress } ) {
+	return <CancelButton onPress={ onPress } text={ __( 'Close' ) } />;
+}
```
I'm not sure if the small text change to "Close" merits an entire component. I'm not sure it provides value. WDYT about leveraging `CancelButton` and passing `text` as a prop when needed instead? 

```jsx
<CancelButton onPress={ () => {} } text={ __( 'Close' ) } />
```

enejb replied:
I am not a big fan of the `text` prop. Mostly because I would expect the text to be passed down with via the children like 
```
<Button onPress={ () => {}  }>{ __( 'Close' ) } ></Button>
```

but in this case, because we are not just rendering all the children but only in special cases. (only on iOS) having a prop of `text` could work well and we can make the component a bit more generic. and remove the need for the different components. 

Another way we would accomplish the same thing would be if a had a specific some sort of `type` and then pass in the different ones. ( 'close', 'back', 'close' ) this would also help us set different accessibility hints in the future. 
I am not a fan of this approach mostly since it creates type is a bit generic term and we end up in a place where we don't really know all the different types.

enejb replied:
Coming back to think I think once we have more specific accessibility Hint and specific content. 
I think it would make sense to have a block for each of the different "BackButtons" s. 
Also, the code would read better and we could avoid things like 
```
<BottomSheet.Header.CancelButton onPress={ close } text={ __( 'Close' ) } />
```

mkevins replied:
I don't have strong opinions about the granularity of the components in this case, but I wanted to mention that there may be cases where "Close" "Cancel" and "Back" are use somewhat inconsistently. The componentization in this PR might also be a good opportunity to clarify the intended semantics of these.

dcalhoun replied:
> I am not a big fan of the `text` prop. Mostly because I would expect the text to be passed down with via the children [...]
>
> but in this case, because we are not just rendering all the children but only in special cases. (only on iOS) having a prop of `text` could work well and we can make the component a bit more generic. and remove the need for the different components.

Understandable. I think relying upon `children` over `text` generally makes sense and should be the default. In reality, the fact that the text is conditionally rendered doesn't prohibit us from setting it with `children`. We could still pass it as `children`.

That said, would it be better to communicate the conditional rendering via the property name? E.g. `iosText`.

```jsx
<BottomSheet.Header.CancelButton iosText={ __('Close') } />
```

> Another way we would accomplish the same thing would be if a had a specific some sort of `type` and then pass in the different ones. ( 'close', 'back', 'close' ) this would also help us set different accessibility hints in the future.
> I am not a fan of this approach mostly since it creates type is a bit generic term and we end up in a place where we don't really know all the different types.

I agree that moving to an abstraction where there is a "type" prop is likely too far an abstraction (also `CancelButton`, `BackButton`, `CloseButton` is sort of already the "types"). The original limitation I ran into with the Help section refactor was that I could not pass custom text for the one-off "Cancel" text needed. 

We could create "types" including one for the "Cancel" scenario, but it sort of puts us right back into the original limitation I hit. What if we encounter a new context where we need "Dismiss" text? If we encounter a new context, do we want to add another "type" or pass the required text via a freeform `text` prop? WDYT? 

> Coming back to think I think once we have more specific accessibility Hint and specific content.
> I think it would make sense to have a block for each of the different "BackButtons" s.

> I don't have strong opinions about the granularity of the components in this case, but I wanted to mention that there may be cases where "Close" "Cancel" and "Back" are use somewhat inconsistently. The componentization in this PR might also be a good opportunity to clarify the intended semantics of these.

The last two quotes are from each of @enejb and @mkevins separately, but I _think_ you are both saying the same thing — that you might prefer to keep `CloseButton`. If I am interpreting that correctly, that is fine with me. My original question of its existence is loosely held. I understand how creating it as a separate component to mange the text and any other accessibility properties provides value. 👍🏻

dcalhoun commented on packages/editor/src/components/editor-help/index.native.js:
```diff
@@ -76,16 +81,12 @@ function EditorHelpTopics( { close, isVisible, onClose } ) {
-							}
-						/>
+						<BottomSheet.Header>
+							<BottomSheet.Header.CloseButton onPress={ close } />
```
As questioned elsewhere, I think we may be better off passing the desired text here. 

```suggestion
							<BottomSheet.Header.CancelButton onPress={ close } text={ __( 'Close' ) } />
```

enejb replied:
I think you mean to write 
```
<BottomSheet.Header.CancelButton onPress={ close } text={ __( 'Close' ) } />
```
Looking at the code I am not sure it is an improvement. Mostly since it creates confusion. 
Maybe we should change it to? 
```
<BottomSheet.Header.ReturnButton onPress={ close } text={ __( 'Close' ) } />
```

dcalhoun replied:
> I think you mean to write
> 
> ```
> <BottomSheet.Header.CancelButton onPress={ close } text={ __( 'Close' ) } />
> ```

Ah, yes. I did make a typo with the `text` string. 

> Looking at the code I am not sure it is an improvement. Mostly since it creates confusion.
> Maybe we should change it to?
> 
> ```
> <BottomSheet.Header.ReturnButton onPress={ close } text={ __( 'Close' ) } />
> ```

Confusion between the "CancelButton" component name and "Close" text? Possibly. If you believe renaming `CancelButton` provides more clarity, that works for me. `ReturnButton` or `DismissButton` could potentially work.

dcalhoun replied:
Leaving one additional note that this conversation will be moot if we decide to retain the `CloseButton`, which I am open to doing.

dcalhoun commented on packages/components/src/mobile/bottom-sheet/header/README.md:
```diff
@@ -0,0 +1,91 @@
+# Header
```
nitpick: My personal opinion is the current documentation is somewhat verbose and unfocused. I put together [an alternative example](https://github.com/WordPress/gutenberg/pull/34339) for consideration, but I recognize it is a lot of change and quite opinionated. Please feel free to ignore it entirely or take anything from it you deem valuable. 🙇🏻

enejb replied:
Thanks for the suggestions in the documentation.

dcalhoun commented on packages/components/src/mobile/bottom-sheet/header/styles.native.scss:
```diff
@@ -0,0 +1,71 @@
+	text-align: center;
+	font-weight: 600;
+	font-size: 16px;
+	position: absolute;
```
When I worked on [the refactor for the Help section](https://github.com/WordPress/gutenberg/commit/89ef4b22f3425755d99e1b0407fa742e8cffa733#diff-f67ea543e569f4574566eade99e242d151ad8f09dcc0e7dd4aeb940a893e11da), I attempted to create [the left-center-right layout](https://codepen.io/davidcalhoun/pen/BaRegBQ) while avoiding absolute positioning. I did this largely to reduce the likelihood of UI collisions. E.g. what happens with a really long title? Do you think there is any value in using the flex layout approach instead of absolute positioning?

enejb replied:
The problem that I ran into is that the title component wasn't centred anymore once I removed the empty View component that was making things centred. 
I am not sure how we can make this happen without using absolute positioning. I will see if I can remove the absolute positing by using align-items: "center".

dcalhoun replied:
Makes sense. If I am not mistaken, my CodePen I linked in my previous comment to does have the title centered even when the right button is not present. I used that in the original Help section header. So, I do believe it is possible.

dcalhoun replied:
That said, it may not be worth the effort or complexity. Absolute position may work well for our use case.

dcalhoun commented on packages/components/src/mobile/bottom-sheet/header/styles.native.scss:
```diff
@@ -0,0 +1,71 @@
+	align-items: flex-start;
+	flex: 1;
+	justify-content: center;
+	z-index: 2;
```
If we avoid absolute positioning, it'd also likely negate the need for any `z-index` declarations.

enejb replied:
:+1:

dcalhoun commented on packages/components/src/font-size-picker/index.native.js:
```diff
@@ -84,10 +84,12 @@ function FontSizePicker( {
+					<BottomSheet.Header.Title>
+						{ label }
+					</BottomSheet.Header.Title>
+				</BottomSheet.Header>
```
I have largely seen "compound" components used to share state via React Context. I have not often seen "compound" components created largely for style application, i.e. providing the structural styles needed to keep the header UI consistent. 

I think this use is really smart, and could work well for (1) ensuring the styles are applied consistently and (2) provide the ability to compose the elements as necessary for each context.

enejb replied:
`BottomSheet.Header.Title` 
I am not really sure if this is desired. I kind of like it mostly because it gives you a context where the compone is meant to be used but this same "advantage" makes it harder to reuse the same component. 

For example. 
```
<BottomSheet.Footer>
<Header.Title>Some footer title</Header.Title>
</BottomSheet.Footer>
```
but I guess at that point we could refactor it to look be part of the bottom sheet title component.  But all that could be avoided if we made the component a bottom sheet component at this point. Thought?

mkevins replied:
Just a thought, but maybe "title" is not the best choice here :thinking: imho there should only be _one_ title on a sheet.. so maybe label or heading is a better name?

dcalhoun replied:
I agree with both of your perspectives. I suppose there is a fine line between naming the is too restrictive and naming so generic that it does not communicate intent. 

I do like "heading" instead of "title." If we desire to allow these components to be used in a footer, we could replace "header" with "navbar." 

```jsx
<BottomSheet.NavBar>
  <BottomSheet.NavBar.Heading>A Footer Component</BottomSheet.NavBar.Heading>
</BottomSheet.NavBar>
```

dcalhoun reviewed (commented):
💯 This is really great work! Thank you! 

I think abstractions like this require a lot of thought to get right. So, thank you for spending the time on this and please have grace with the large amounts of comments questions I pose. 😄 I appreciate you putting this together. 

I noticed that the bottom sheet for the Columns block variation selection _does not_ use these components. Is that intentional? I would think we should use these new components there as well for consistency. WDYT?

dcalhoun committed:
Update BottomSheet Header documentation to reduce its size (#34339)

Leverage short, declarative statements when possible.

mkevins commented on packages/components/src/mobile/bottom-sheet/header/README.md:
```diff
@@ -0,0 +1,67 @@
+
+Callback invoked once the button is pressed.
+
+## BottomSheet.Header.CancelButton
+
+Displays a styled button to dismiss a full screen bottom sheet.
+
+### Props
+
+#### onPress
+
+Callback invoked once the button is pressed.
+
+## BottomSheet.Header.CloseButton
+
+Displays a styled button to dismiss a full screen bottom sheet.
+
+### Props
+
+#### onPress
+
+Callback invoked once the button is pressed.
```
Though the description here is the same for both of these components, I'm wondering about their intended semantics, and whether they really are the same. I think that Cancel implies "undoing" some changes that were made, while Close does not imply that. Otoh, what actually happens is entirely dependent on the implementation in the callback. Maybe we can find a way to define the difference between these components in the readme to guide developers that use these components toward a more consistent usage?

dcalhoun replied:
> Otoh, what actually happens is entirely dependent on the implementation in the callback. 

That was my original thought with the documentation for the `onPress` callback. The current description is very much focused on what _does_ occur, not the intended usage.

> Maybe we can find a way to define the difference between these components in the readme to guide developers that use these components toward a more consistent usage?

Counter to the current documentation (I modified) and my aforementioned original thought, I do understand and agree there is value in describing the intent. The description of the components themselves differ somewhat from the `onPress` prop description. E.g. `ApplyButton` states "Displays a styled button to apply settings of bottom sheet controls," which describe the intent to apply the settings. 

If we want to update the documentation to better describe the intent of the components and props, that is fine with me. Apologies if my previous changes removed too much of that, and thus creating more work to add the intent back. 😄

mkevins replied:
> Apologies if my previous changes removed too much of that, and thus creating more work to add the intent back.

No worries! I think it's great to discuss this and I'm happy this is being addressed. The usage of these patterns will certainly grow, so it's valuable to consider many paths. There is a tendency (and maybe a goal?) to make changes to block settings immediately reflected in the block rendering (i.e. no "confirm" or "cancel" needed). I think this can't cover the case for any fullscreen sheets, though, so this may be _at least one_ source of tension. Historically, though, this may have been addressed in a somewhat ad-hoc way, giving rise to some redundancy in naming — but that's just my guess.

In any case, removing truly synonymous components and clarifying the usage of the remaining ones is a great idea. Also, I especially like [the idea mentioned here](https://github.com/WordPress/gutenberg/pull/34309#discussion_r697477701) to add `iosText` as a prop. This accurate signalling provides immediate clarity, whereas the current form feels like it is violating the principle of least surprise, imo.

mkevins reviewed (approved):
Thanks for these changes Enej! I made a few comments / suggestions, but none of them are blockers.

Also, I tested this for the link picker in the format library, and these changes don't seem to introduce any new regressions there. :+1:

enejb committed:
Move Header to NavBar

enejb committed:
Update usage from Header to NavBar

enejb committed:
Remove the NavigationHeader component

enejb committed:
Fix space in the styles

enejb commented:
Thanks for all the feedback and reviews @SiobhyB , @dcalhoun  and @mkevins ! 

In my last update, I did the following:
- Replaced the `CancelButton` and `CloseButton` with a `DismissButton` that defaults to "Cancel" and has an `iosText` prop to accommodate the "close" button case. 
- Updated the Components to be called. `NavBar` and `Heading` instead of Header and Title. So that they are a bit more generic. 
- Update the read me. to add more clarity.

I think things are in pretty good shape. 

I tested this as best as I could on iOS and Android and it seems it works now as before. 

Things to get done in different PRs. 
[ ] Get clarification when it comes to the Touch Feedback. 
[ ] Update the Variation control to use this new NavBar Component. (Columns Block) 
[ ] Update the accessibilityLabels so that they make more sense for the different Dismiss / Back components.

dcalhoun reviewed (approved):
LGTM. Well done! I retested on an iPhone 12 mini simulator and Android Pixel 5 emulator.
